atc/worker/worker.go

```diff
@@ -207,6 +211,72 @@ func (worker *gardenWorker) LookupVolume(logger lager.Logger, handle string) (Vo
 	return worker.volumeClient.LookupVolume(logger, handle)
 }
 
+var sensitiveKeys = []string{"password", "key", "token"}
```

Comment:
Ah, so this must have been one of the motivations for only checking certain fields. A blocklist approach is indeed kinda risky.

I wonder: if this check is performed before credentials are resolved (which I *think* may be the case - Concourse tries to evaluate creds as late as possible), do you think we could just pass the whole payload along? This won't help for people who aren't using credential management, but, well... they probably should be. There may be some other case I'm missing though.

More generally: is the network traffic to OPA expected to be secure and trusted? If all else fails, how worried are we if credentials are sent along?

Scope creep, not needed for this PR: I wonder if we should introduce another field alongside `source:` for storing sensitive fields so that in situations like this we can simply not send them along. Being able to distinguish which config "identifies" the resource and which config is necessary to access it would likely be helpful in other scenarios within Concourse.

Reply:
> I wonder: if this check is performed before credentials are resolved (which I _think_ may be the case - Concourse tries to evaluate creds as late as possible), do you think we could just pass the whole payload along? This won't help for people who aren't using credential management, but, well... they probably should be. There may be some other case I'm missing though.
> 

I believe this check is done after var interpolation. Because var evaluation happens in beginning of a step, and then a step tries to launch a container.

> More generally: is the network traffic to OPA expected to be secure and trusted? If all else fails, how worried are we if credentials are sent along?
> 

Traffic between concourse and opa could be either secure or insecure, depends on how it's setup. But, policy checking might go to concourse debug log, so secrets might be printed to concourse logs, and concourse logs are not supposed to be secrets. In our clusters, concourse logs are redirected to log-insight, and everyone has access to log-insight. Similarly at OPA side, an OPA admin may turn on debug log, then all policy check requests are printed to OPA logs. That's the main concern I have.

> Scope creep, not needed for this PR: I wonder if we should introduce another field alongside `source:` for storing sensitive fields so that in situations like this we can simply not send them along. Being able to distinguish which config "identifies" the resource and which config is necessary to access it would likely be helpful in other scenarios within Concourse.

That's a good idea, but that would be a breaking change. Even if such a field is introduced, I guess we have to be back-compatible `source` for existing pipelines for a while, thus the new field doesn't help in short term. But for long term, I'm totally on this idea.

I just pushed a new commit, that addressed the other two comments but this one. Let me think more on this today.

Reply:
In the latest commit, I leverage secret redaction via build delegate to redact secrets in image source. Please see what you feel. I don't want to redact the entire message sent to OPA, because I have seen many users store non-secret in Vault, which causes unintentional redactions, and might lead to policy checking rejected.